Plasma flower MOD core from the lavaland patch of eden ruin

Momo8289 opened the pull request:
## About The Pull Request
This PR gives a small update to the patch of eden lavaland ruin, and adds a new item that can be used as an upgrade to the plasma mod core. The idea around this item is that it's a flower that has absorbed all the plasma and bad-vibes in the small area around the ruin, leaving an area full of plants, and a flower full of energy. This flower can be taken from the oasis and turned into a functional MOD core using some wires. As a MOD core, it is the same as the plasma mod core, but has a higher energy capacity, being equivalent to a super power cell. It also gives you a cool pollen effect, and spawns butterflies around you while your suit is active. These butterflies disappear when they get too far away, or if the suit deactivates. They also don't leave corpses, so they shouldn't cause too much clutter.
![image](https://github.com/tgstation/tgstation/assets/42454181/45d67bd6-d001-4604-9814-22f32b8d63ef)


https://github.com/tgstation/tgstation/assets/42454181/22610231-4b98-440d-bf99-ae5b88644de8




## Why It's Good For The Game
Secondary, more concrete reason: The patch of eden ruin is overall pretty boring for what it could be, and adding a cool thing like this could spice it up imo. I also think the plasma mod core is cool, and so making it possible to get it outside of cannibalizing a mining MODsuit would be nice.

Now, the main motivation behind this PR, and my future PRs, is that I want to improve the interactions between Robotics and the rest of the station, starting with miners. By and large, interacting with robotics as another department usually sucks. Something I believe to be a major contributor to this is the fact that there's nothing for the rest of the crew to really help robotics with. Robotics gains basically nothing from interacting with the rest of the crew. The idea is that a miner finds this core on lavaland, brings it to robotics, and robotics gets a cool new mod core. Obviously the miners can use the core themselves, but in my experience, most miners don't really use the MODsuit. I'm trying to appeal to the basic feeling of "This person brought me something, I'll do something for them". Obviously this is a pretty minor step, but hey, gotta start somewhere.
## Changelog
:cl:
add: Updated the Patch of Eden lavaland ruin with a new item
add: Added the Plasma flower item, which can be used as an upgraded plasma MOD core
/:cl:

Momo8289 committed:
Added plasma flower mod core

Momo8289 committed:
Added plasma flower item and sprite
Can be turned into the mod core with wire coil

Momo8289 committed:
Updated patch of eden to include the item
And some ruins

Momo8289 committed:
fixed last commit, was weird from fulp code

Momo8289 committed:
Merge branch 'plasma-flower' of github.com:Momo8289/tgstation into plasma-flower

Momo8289 committed:
hopefully fixes conflicts

Fikou commented on code/modules/mod/mod_construction.dm:
```diff
@@ -51,6 +51,33 @@
+	name = "Plasma flower"
+	icon_state = "mode-plasma-flower"
+	desc = "A strange flower from the desolate wastes of lavaland. It pulses with a bright purple glow.  \
+	Its shape is remarkably similar to that of a MOD core."
```
```suggestion
		Its shape is remarkably similar to that of a MOD core."
```

Fikou commented on code/modules/mod/mod_construction.dm:
```diff
@@ -51,6 +51,33 @@
 	qdel(src)
 
+/obj/item/mod/construction/lavalandcore
+	name = "Plasma flower"
```
```suggestion
	name = "plasma flower"
```

Fikou commented on code/modules/mod/mod_construction.dm:
```diff
@@ -51,6 +51,33 @@
+	. += span_notice("You could probably attach some <b>wires</b> to it...")
+
+/obj/item/mod/construction/lavalandcore/attackby(obj/item/P, mob/user, params)
+	if(istype(P, /obj/item/stack/cable_coil))
```
use early returns

Fikou commented on code/modules/mod/mod_construction.dm:
```diff
@@ -51,6 +51,33 @@
+	. = ..()
+	. += span_notice("You could probably attach some <b>wires</b> to it...")
+
+/obj/item/mod/construction/lavalandcore/attackby(obj/item/P, mob/user, params)
```
use descriptive argument names, not single letters

Fikou commented on code/modules/mod/mod_construction.dm:
```diff
@@ -51,6 +51,33 @@
+		if(!P.use_tool(src, user, 5 SECONDS, amount = 2, volume = 30))
+			return
+
+		to_chat(user, span_notice("You add the wires to the core"))
```
```suggestion
		to_chat(user, span_notice("You add the wires to the core."))
```

Fikou commented on code/modules/mod/mod_core.dm:
```diff
@@ -355,3 +355,17 @@
 	return TRUE
+
+/obj/item/mod/core/plasma/lavaland
+	name = "Plasma flower MOD core"
```
```suggestion
	name = "plasma flower MOD core"
```

Fikou commented on code/modules/mod/mod_core.dm:
```diff
@@ -355,3 +355,17 @@
+	name = "Plasma flower MOD core"
+	icon_state = "mod-core-plasma"
+	desc = "A strange flower from the desolate wastes of lavaland. It pulses with a strange purple glow.  \
+	The wires coming out of it could be hooked into a MODsuit."
```
```suggestion
		The wires coming out of it could be hooked into a MODsuit."
```

Momo8289 committed:
Update code/modules/mod/mod_core.dm from suggestion

Co-authored-by: Fikou <23585223+Fikou@users.noreply.github.com>

Momo8289 committed:
Updated var name to be more descriptive

Momo8289 committed:
Merge branch 'plasma-flower' of github.com:Momo8289/tgstation into plasma-flower

Momo8289 committed:
Updated mod core name to fit other mod cores

Momo8289 committed:
Update to use early return

Momo8289 reviewed (commented):
Yeah mixed up my commit name here, sorry. Actually just updated some text.

Momo8289 committed:
Actually updates MOD core name to fit other cores

Momo8289 committed:
Updated plasma flower name

Momo8289 committed:
Updated desc

Momo8289 committed:
Forgot to update var name

carlarctg commented:
Interlinking robotics and modsuits with other parts of the game such as ruins, arcade loot, bepis, is something that is critically underdone right now, this is a cool way to do so, though it'd be nice if it did *something* besides being a slightly better core. For example, spawning butterflies that delete themselves around you, flowers on the ground (ash flowers if on lavaland), etc.

Momo8289 commented:
> For example, spawning butterflies that delete themselves around you, flowers on the ground (ash flowers if on lavaland), a flower on your modsuit head, etc

Ooh, that would be cool. I'll see what I can do, but I'm pretty new to DM so no promises.

carlarctg commented:
> > For example, spawning butterflies that delete themselves around you, flowers on the ground (ash flowers if on lavaland), a flower on your modsuit head, etc
> 
> Ooh, that would be cool. I'll see what I can do, but I'm pretty new to DM so no promises.

Are you on the discord? I can help you with the code if so.

Momo8289 commented:
> Are you on the discord? I can help you with the code if so.

I am now! Momo#9817. I'm mostly figuring stuff out on my own, but I do have a couple of questions so help would be great!

Momo8289 committed:
Added new butterfly to be spawned by the core

Momo8289 committed:
The core now spawns the butterflied while active

Momo8289 committed:
Updated sprite for plasma flower

Momo8289 committed:
updated icon

Momo8289 committed:
child_despawned is now called on destroy

Momo8289 committed:
reduced the number of butterflies

Momo8289 committed:
Added pollen effect

Momo8289 committed:
Will not spawn when over/at cap

Momo8289 committed:
Merge branch 'master' into plasma-flower

Momo8289 committed:
checks for source before calling child_despawned

Momo8289 committed:
now cleans up properly

Momo8289 committed:
Fixed cleanup again

Momo8289 committed:
Merge branch 'master' into plasma-flower

Momo8289 committed:
Butterflies now fade away when they despawn

carlarctg commented on code/modules/mob/living/basic/vermin/butterfly.dm:
```diff
@@ -44,3 +44,55 @@
+
+/mob/living/basic/butterfly/lavaland/temporary/proc/fadeout()
+	animate(src, alpha = 0, 1 SECONDS)
+	addtimer(CALLBACK(src, TYPE_PROC_REF(/mob/living/basic/butterfly/lavaland/temporary, despawn)), 1 SECONDS)
```
you dont need to do this (i think), just addtimer for qdel, same thing

Jacquerel commented on code/modules/mob/living/basic/vermin/butterfly.dm:
```diff
@@ -44,3 +44,55 @@
+			deltimer(despawn_timer)
+
+/mob/living/basic/butterfly/lavaland/temporary/proc/should_despawn()
+	if(!source.mod.active)
```
This should also use a signal, register to `COMSIG_MODULE_DEACTIVATED` and trigger the deaths of all butterflies at that point.

Jacquerel commented on code/modules/mob/living/basic/vermin/butterfly.dm:
```diff
@@ -44,3 +44,55 @@
+
+/mob/living/basic/butterfly/lavaland/temporary
+	name = "strange butterfly"
+	var/obj/item/mod/core/plasma/lavaland/source = NONE
```
If you switch to using signals for everything except distance tracking, you can just cast this as an `atom` instead and then have temporary butterflies which live within a range of any atom.

Jacquerel commented on code/modules/mob/living/basic/vermin/butterfly.dm:
```diff
@@ -44,3 +44,55 @@
+/mob/living/basic/butterfly/lavaland/temporary/process()
+	if(should_despawn())
+		will_be_destroyed = TRUE
+		despawn_timer = addtimer(CALLBACK(src, TYPE_PROC_REF(/mob/living/basic/butterfly/lavaland/temporary, fadeout)), 5 SECONDS, TIMER_STOPPABLE)
```
Does this not call repeatedly and reset the timer? You don't check if `will_be_destroyed` has already toggled to true.

Jacquerel commented on code/modules/mob/living/basic/vermin/butterfly.dm:
```diff
@@ -44,3 +44,55 @@
+
+/mob/living/basic/butterfly/lavaland/temporary/death(gibbed)
+	. = ..()
+	qdel(src)
```
This is covered by adding `basic_mob_flags = DEL_ON_DEATH`

Jacquerel commented on code/modules/mod/mod_core.dm:
```diff
@@ -355,3 +355,70 @@
+	light_color = "#cc00cc"
+	light_range = 2
+
+	var/mob/living/mob = /mob/living/basic/butterfly/lavaland/temporary
```
These also need documenting comments, and this should probably have a more specific name.

Jacquerel commented on code/modules/mod/mod_core.dm:
```diff
@@ -355,3 +355,70 @@
+/obj/item/mod/core/plasma/lavaland/Destroy()
+	for(var/mob/child in children)
+		qdel(child)
+	if(mod)
+		if(mod.wearer)
```
```suggestion
	if(mod?.wearer)
```

Jacquerel commented on code/modules/mod/mod_core.dm:
```diff
@@ -355,3 +355,70 @@
+/obj/item/mod/core/plasma/lavaland/process()
+	if(can_spawn)
+		can_spawn = FALSE
+		addtimer(CALLBACK(src, TYPE_PROC_REF(/obj/item/mod/core/plasma/lavaland, spawn_mob)), 5 SECONDS)
```
This should probably be replaced with a cooldown, check out `/datum/component/spawner` for an example.
Alternately, can you just attach a `spawner` component to this in some way?

Jacquerel commented on code/modules/mod/mod_core.dm:
```diff
@@ -355,3 +355,70 @@
+		addtimer(CALLBACK(src, TYPE_PROC_REF(/obj/item/mod/core/plasma/lavaland, spawn_mob)), 5 SECONDS)
+
+/obj/item/mod/core/plasma/lavaland/proc/child_despawned()
+	spawned--
```
Why handle this count separately instead of just using the length of the `children` list?

Jacquerel commented on code/modules/mod/mod_core.dm:
```diff
@@ -355,3 +355,70 @@
+		can_spawn = TRUE
+
+/obj/item/mod/core/plasma/lavaland/proc/spawn_mob()
+	SIGNAL_HANDLER
```
This isn't a signal handler.

Jacquerel reviewed (changes requested):
This is definitely an interesting add, but I think you'll want to make more use of signals here, and maybe even some existing components we already have to avoid reinventing the wheel.

Momo8289 committed:
Minor code fixes

Momo8289 committed:
butterflies now use DEL_ON_DEATH

Momo8289 committed:
Added signal for when a mob spawner spawns a mob

Momo8289 committed:
The mod core now uses /datum/component/spawner
to spawn mobs

Momo8289 committed:
Butterflies are now properly deleted
 when the mod shuts down

Momo8289 committed:
Butterfly documentation

Momo8289 committed:
Merge branch 'master' into plasma-flower

Momo8289 committed:
removed a var used for debugging

Momo8289 committed:
Added flower trail while suit is active

Momo8289 commented:
> This is definitely an interesting add, but I think you'll want to make more use of signals here, and maybe even some existing components we already have to avoid reinventing the wheel.

Thanks for pointing out `/datum/component/spawner`, it was pretty much exactly what I was after lol. I think I've addressed most of your feedback, although things have changed a fair bit so I don't doubt there will be more.

Momo8289 committed:
Merge branch 'master' into plasma-flower

JohnFulpWillard commented on code/modules/mob/living/basic/vermin/butterfly.dm:
```diff
@@ -44,3 +44,48 @@
+			will_be_destroyed = FALSE
+			deltimer(despawn_timer)
+
+/mob/living/basic/butterfly/lavaland/temporary/proc/should_despawn()
```
Can you add documentation to what this proc does

JohnFulpWillard commented on code/modules/mob/living/basic/vermin/butterfly.dm:
```diff
@@ -44,3 +44,48 @@
+
+/mob/living/basic/butterfly/lavaland/temporary/Destroy()
+	STOP_PROCESSING(SSprocessing, src)
+	. = ..()
```
```suggestion
	return ..()
```

JohnFulpWillard commented on code/modules/mob/living/basic/vermin/butterfly.dm:
```diff
@@ -44,3 +44,48 @@
+			return
+		will_be_destroyed = TRUE
+		despawn_timer = addtimer(CALLBACK(src, TYPE_PROC_REF(/mob/living/basic/butterfly/lavaland/temporary, fadeout)), 5 SECONDS, TIMER_STOPPABLE)
+	else
```
You can early return ``should_despawn()`` instead of using an else, if you want. This proc doesn't do anything else so it would look nicer

JohnFulpWillard commented on code/modules/mod/mod_core.dm:
```diff
@@ -355,3 +355,71 @@
+	light_color = "#cc00cc"
+	light_range = 2
+
+	/// The mob to be spawned by the core
+	var/mob/living/spawned_mob = /mob/living/basic/butterfly/lavaland/temporary
+	/// Max number of mobs it can spawn
+	var/max_spawns = 3
+	var/can_spawn = TRUE
+	/// Mob spawner for the core
+	var/datum/component/spawner/mob_spawner
```
Newly defined vars should be at the bottom, in this case ``maxcharge`` and ``charge`` should be above these.

JohnFulpWillard commented on code/modules/mod/mod_core.dm:
```diff
@@ -355,3 +355,71 @@
+	var/mob/living/spawned_mob = /mob/living/basic/butterfly/lavaland/temporary
+	/// Max number of mobs it can spawn
+	var/max_spawns = 3
+	var/can_spawn = TRUE
```
Can you add a comment to this

JohnFulpWillard commented on code/modules/mod/mod_core.dm:
```diff
@@ -355,3 +355,71 @@
+	light_range = 2
+
+	/// The mob to be spawned by the core
+	var/mob/living/spawned_mob = /mob/living/basic/butterfly/lavaland/temporary
```
If you want to be more descriptive, this can be ``spawned_mob_type``

JohnFulpWillard commented on code/modules/mod/mod_core.dm:
```diff
@@ -355,3 +355,71 @@
+/obj/item/mod/core/plasma/lavaland/Destroy()
+	if(mod?.wearer)
+		mod.wearer.particles = null
+	. = ..()
```
```suggestion
	return ..()
```

JohnFulpWillard commented on code/modules/mod/mod_core.dm:
```diff
@@ -355,3 +355,71 @@
+
+/obj/item/mod/core/plasma/lavaland/proc/flowas(mob/living/wearer)
+	SIGNAL_HANDLER
+	var/static/list/possible_flower_types = list(/obj/structure/flora/bush/lavendergrass/style_random, /obj/structure/flora/bush/flowers_yw/style_random,       /obj/structure/flora/bush/flowers_br/style_random, /obj/structure/flora/bush/flowers_pp/style_random)
```
Multi-line this please

Momo8289 committed:
Minor code refactors and documentation

Momo8289 committed:
Removed unused var

Momo8289 committed:
Minor code fixes

JohnFulpWillard commented on code/modules/mod/mod_core.dm:
```diff
@@ -355,3 +355,77 @@
+		/obj/structure/flora/bush/lavendergrass/style_random,
+		/obj/structure/flora/bush/flowers_yw/style_random,
+		/obj/structure/flora/bush/flowers_br/style_random,
+		/obj/structure/flora/bush/flowers_pp/style_random
```
```suggestion
		/obj/structure/flora/bush/flowers_pp/style_random,
```

Momo8289 replied:
Will do. In general, should I leave trailing commas for stuff like this?

JohnFulpWillard commented on code/modules/mod/mod_core.dm:
```diff
@@ -355,3 +355,77 @@
+		/obj/structure/flora/bush/flowers_br/style_random,
+		/obj/structure/flora/bush/flowers_pp/style_random
+	)
+	var/chosen_type
+	if(!chosen_type || prob(5))
+		chosen_type = pick(possible_flower_types)
```
I don't understand this bit, ``chosen_type`` will never be already set here, why do we have it? And what's the prob(5) supposed to do?

JohnFulpWillard commented on code/modules/mob/living/basic/vermin/butterfly.dm:
```diff
@@ -44,3 +44,50 @@
+		return TRUE
+	return FALSE
+
+/mob/living/basic/butterfly/lavaland/temporary/proc/fadeout()
```
Forgot to say but this should also be documented

Momo8289 committed:
comments

Momo8289 committed:
Code cleanup

JohnFulpWillard commented on code/modules/mob/living/basic/vermin/butterfly.dm:
```diff
@@ -44,3 +44,52 @@
+	return FALSE
+
+/// Fade the butterfly out before deleting it.
+// Looks much better than it just blipping out of existence
```
```suggestion
/// Looks much better than it just blipping out of existence
```

Momo8289 committed:
Update code/modules/mob/living/basic/vermin/butterfly.dm

Co-authored-by: John Willard <53777086+JohnFulpWillard@users.noreply.github.com>
